File: app/models/lab.py
"""
InfecSure — Pydantic Models: Laboratory Results
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional

from pydantic import BaseModel


class SpecimenType(str, Enum):
    BLOOD = "blood"
    URINE = "urine"
    SPUTUM = "sputum"
    WOUND_SWAB = "wound_swab"
    CSF = "csf"
    STOOL = "stool"
    OTHER = "other"


class AnomalyFlag(BaseModel):
    is_anomaly: bool
    z_score: float
    message: Optional[str] = None
    severity: Optional[str] = None  # "warning" | "critical"


class LabResultCreate(BaseModel):
    ward_id: str
    pathogen_id: str
    pathogen_name: str                  # Denormalized for fast reads
    specimen_type: SpecimenType
    result_date: datetime
    test_result: Optional[str] = None
    colony_count: Optional[int] = None
    resistance_profile: Optional[list[str]] = []  # e.g. ["MRSA", "VRE"]
    antibiotic_sensitivity: Optional[dict] = {}   # {"ampicillin": "R", "ciprofloxacin": "S"}
    patient_ward_location: Optional[str] = None   # ward/bed reference (masked for staff)
    notes: Optional[str] = None


class LabResult(LabResultCreate):
    result_id: str
    entered_by_uid: str
    entered_by_name: str
    status: str = "pending"
    anomaly: Optional[AnomalyFlag] = None
    created_at: Optional[datetime] = None


class LabResultPublic(BaseModel):
    """Masked version for Staff/Doctor role — no patient identifiers."""
    result_id: str
    ward_id: str
    pathogen_name: str
    specimen_type: SpecimenType
    result_date: datetime
    test_result: Optional[str] = None
    status: str = "pending"
    anomaly: Optional[AnomalyFlag] = None
    created_at: Optional[datetime] = None


class LabSlipOCRRequest(BaseModel):
    image_base64: str


class LabSlipOCRResponse(BaseModel):
    raw_text: str
    extracted_fields: dict[str, Any]
    tokens: list[dict[str, Any]] = []
    low_confidence_count: int = 0
    engine: str = "easyocr"

File: InfecSure_Frontend/src/api/lab.ts
import { api } from "./client";
import type { LabResult } from "../types";

export type LabResultPayload = {
  ward_id: string;
  pathogen_id: string;
  pathogen_name: string;
  specimen_type: string;
  result_date: string;
  test_result?: "positive" | "negative";
  colony_count?: number;
  resistance_profile?: string[];
  antibiotic_sensitivity?: Record<string, string>;
  patient_ward_location?: string;
  notes?: string;
};

export type LabSlipOCRResponse = {
  raw_text: string;
  extracted_fields: Partial<LabResultPayload> & {
    test_result?: "positive" | "negative";
  };
  tokens: Array<Record<string, unknown>>;
  low_confidence_count: number;
  engine: "google_vision" | "easyocr" | string;
};

export async function createLabResult(payload: LabResultPayload) {
  const { data } = await api.post("/lab-results/", payload);
  return data;
}

export async function listLabResults(wardId?: string) {
  const { data } = await api.get<LabResult[]>("/lab-results/", { params: wardId ? { ward_id: wardId } : undefined });
  return data;
}

export async function scanLabSlip(imageBase64: string) {
  const { data } = await api.post<LabSlipOCRResponse>("/lab-results/scan-slip", { image_base64: imageBase64 });
  return data;
}

export async function getLabResult(resultId: string) {
  const { data } = await api.get<LabResult>(`/lab-results/${resultId}`);
  return data;
}
